lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -172,15 +169,44 @@ SILInstruction *SILCombiner::visitSwitchEnumAddrInst(SwitchEnumAddrInst *SEAI) {
   //     ->
   //   %value = load %ptr
   //   switch_enum %value
-  SmallVector<std::pair<EnumElementDecl*, SILBasicBlock*>, 8> Cases;
-  for (int i = 0, e = SEAI->getNumCases(); i < e; ++i)
+  //
+  // If we are using ownership, we perform a load_borrow right before the new
+  // switch_enum and end the borrow scope right afterwards.
+  Builder.setCurrentDebugScope(SEAI->getDebugScope());
+  SmallVector<std::pair<EnumElementDecl *, SILBasicBlock *>, 8> Cases;
+  for (int i : range(SEAI->getNumCases())) {
     Cases.push_back(SEAI->getCase(i));
+  }
 
-  Builder.setCurrentDebugScope(SEAI->getDebugScope());
   SILBasicBlock *Default = SEAI->hasDefault() ? SEAI->getDefaultBB() : nullptr;
-  LoadInst *EnumVal = Builder.createLoad(SEAI->getLoc(), Addr,
-                                         LoadOwnershipQualifier::Unqualified);
-  Builder.createSwitchEnum(SEAI->getLoc(), EnumVal, Default, Cases);
+  SILValue EnumVal = Builder.emitLoadBorrowOperation(SEAI->getLoc(), Addr);
```

Comment:
It took me a while to figure out how this is safe. It would help to comment that this only creates a trivial load-borrow scope to cover the switch_enum terminator and its block argument results, so nothing can write to memory within this scope.

Reply:
I'll add to that.

Reply:
I looked, I do have a comment a line or two above:

```
  // Promote switch_enum_addr to switch_enum if the enum is loadable.                                                                    
  //   switch_enum_addr %ptr : $*Optional<SomeClass>, case ...                                                                           
  //     ->                                                                                                                              
  //   %value = load %ptr                                                                                                                
  //   switch_enum %value                                                                                                                
  //                                                                                                                                     
  // If we are using ownership, we perform a load_borrow right before the new                                                            
  // switch_enum and end the borrow scope right afterwards.
```